### PR TITLE
Add new heads subscribption

### PR DESCRIPTION
**Pull request by vldm**

### CODE REVIEW

File: core/src/rpc_pubsub.rs [hunk 1]
```diff
@@ -536,6 +554,46 @@ impl RpcSolPubSub for RpcSolPubSubImpl {
...
+        } else {
+            error!(
+                "eth_subscribe: Trying to subscribe to unsupported topic: {}",
+                topic
+            )
+        }
+    }
+
+    // Unsubscribe from new heads notifications.
+    fn eth_unsubscribe(&self, _meta: Option<Self::Metadata>, id: SubscriptionId) -> Result<bool> {
```
Comment (hrls) [inline core/src/rpc_pubsub.rs:585]: Maybe `Result<()>` till `Ok(false)` isn't there.

File: core/src/rpc_subscriptions.rs [hunk 2]
```diff
@@ -1014,6 +1022,36 @@ impl RpcSubscriptions {
...
+    pub fn remove_evm_new_blocks(&self, id: &SubscriptionId) -> bool {
+        let mut subscriptions = self.subscriptions.evm_block_subscriptions.write().unwrap();
+        subscriptions.remove(id).is_some()
+    }
+
+    pub fn notify_evm_block(&self, new_head: evm_state::Block) {
+        let transactions = new_head
+            .transactions
+            .iter()
+            .map(|(k, _)| *k)
```
Comment (hrls) [inline core/src/rpc_subscriptions.rs:1044]: `.keys().copied()`?

File: core/src/evm_rpc_impl/mod.rs [hunk 3]
```diff
@@ -163,30 +175,11 @@ impl ChainMockERPC for ChainMockErpcImpl {
...
 
-        let result = RPCBlock {
-            number: U256::from(block.header.block_number).into(),
-            hash: block_hash.into(),
-            parent_hash: parent_hash.into(),
-            size: 0x100.into(),
-            gas_limit: Hex(block.header.gas_limit.into()),
-            gas_used: Hex(block.header.gas_used.into()),
-            timestamp: Hex(block.header.timestamp),
+        Ok(Some(RPCBlock::new_from_head(
```
Comment (hrls) [inline core/src/evm_rpc_impl/mod.rs:178]: Probably `RPCBlock::from_block(block)` constructor would be more suitable.

File: core/src/rpc_pubsub.rs (hunk 1 above)
Comment (vldm) [reply core/src/rpc_pubsub.rs:585]: https://geth.ethereum.org/docs/rpc/pubsub

File: core/src/rpc_subscriptions.rs [hunk 4]
```diff
@@ -1014,6 +1022,36 @@ impl RpcSubscriptions {
...
+        &self,
+        sub_id: SubscriptionId,
+        subscriber: Subscriber<evm_rpc::RPCBlock>,
+    ) {
+        let sink = subscriber.assign_id(sub_id.clone()).unwrap();
+        let mut subscriptions = self.subscriptions.evm_block_subscriptions.write().unwrap();
+        subscriptions.insert(sub_id, sink);
+    }
+
+    pub fn remove_evm_new_blocks(&self, id: &SubscriptionId) -> bool {
```
Comment (hrls) [inline core/src/rpc_subscriptions.rs:1035]: Function name is too short. Actual effects is in removing subscriptions, not `evm_new_blocks`.

File: core/src/rpc_subscriptions.rs [hunk 5]
```diff
@@ -1014,6 +1022,36 @@ impl RpcSubscriptions {
         });
     }
 
+    pub fn add_evm_new_blocks(
```
Comment (hrls) [inline core/src/rpc_subscriptions.rs:1025]: Probably `fn add_evm_block_subscription`

File: ledger/src/blockstore.rs [hunk 6]
```diff
@@ -2473,7 +2476,9 @@ impl Blockstore {
         _block_slot: Slot,
         block: &evm::BlockHeader,
     ) -> Result<()> {
-        self.evm_blocks_cf.put(block.block_number, block)
+        self.evm_blocks_cf.put(block.block_number, block)?;
+        error!("Printing block_hash = {:?}", block.hash());
```
Comment (hrls) [inline ledger/src/blockstore.rs:2480]: `info!`?

File: ledger/src/blockstore.rs [hunk 7]
```diff
@@ -2483,6 +2488,26 @@ impl Blockstore {
...
+    ) -> Result<()> {
+        let mut w_active_transaction_status_index =
+            self.active_transaction_status_index.write().unwrap();
+        let primary_index = self.get_primary_index(slot, &mut w_active_transaction_status_index)?;
+        self.evm_blocks_by_hash_cf.put((primary_index, hash), &id)
+    }
+    pub fn read_evm_block_id_by_hash(&self, hash: H256) -> Result<Option<evm_state::BlockNum>> {
+        let result = self.evm_blocks_by_hash_cf.get((0, hash))?;
+        if result.is_none() {
+            Ok(self.evm_blocks_by_hash_cf.get((1, hash))?)
```
Comment (hrls) [inline ledger/src/blockstore.rs:2505]: How this pair works?

**Pull request merged**